webapp/src/utils/jira_issue_metadata.tsx

```diff
@@ -227,21 +228,38 @@ export function getCustomFieldFiltersForProjects(metadata: IssueMetadata | null,
         } as FilterField);
     }
 
-    if (issueStatuses) {
+    // Filtering out the statuses on th basis of selected issue types
+    const issueTypesWithStatuses = metadata && metadata.issue_types;
+    const keys: string[] = [];
+    const statuses: Status[] = [];
+    if (issueTypesWithStatuses) {
+        issueTypesWithStatuses.forEach((element: IssueTypeWithStatuses) => {
+            if (issueTypes.includes(element.id) || !issueTypes.length) {
+                element.statuses.forEach((status) => {
+                    if (!keys.includes(status.id)) {
+                        keys.push(status.id);
+                        statuses.push(status);
+                    }
+                });
+            }
+        });
+    }
+
```

Comment:
```suggestion
const issueTypesWithStatuses = metadata?.issue_types;
const keys = new Set<string>();
const statuses: Status[] = [];

if (issueTypesWithStatuses) {
    for (const element of issueTypesWithStatuses) {
        if (issueTypes.includes(element.id) || !issueTypes.length) {
            for (const status of element.statuses) {
                if (!keys.has(status.id)) {
                    keys.add(status.id);
                    statuses.push(status);
                }
            }
        }
    }
}
```
Always looks for the scope of optimization as well
- Replace forEach with for...of loops: This change may lead to a small performance improvement as for...of is generally more performant than forEach.

- Change keys to a Set: This will ensure that keys.includes(status.id) operates in O(1) time complexity instead of O(n) as it would with an array.